Fixes issues with filter tree, moves query basic to use filterTree directly

andrewkfiedler opened the pull request:
See commits for breakdown.  Big part of this was enforcing classes, which involved making sure typescript didn't allow using the spread operator (which results in a plain object) without feeding that into a new instance of the class.

Also optimized some bits for performance since depending on the filter tree directly caused more frequent updates to it.  Notably there is a change in the home component, and we now memo the query settings.

Testing:
Go wild testing everything on the basic page, then test going to the advanced and ensuring it translates appropriately.

In advanced, test out grouping, negation, etc.

andrewkfiedler committed:
Update filter builder to enforce class / immutability

 - Updates filter structure to make the classes reject spread operations (this turns them into plain objects).
 - Updates filter structure to make the classes immutable for the most part.
 - Removes non working tests

andrewkfiedler committed:
Updates query basic to use filterTree directly, like query advanced now does

 - Various updates to have query basic depend directly on the filter tree.
 - Update to a function in home to debounce it, because with all the frequent changes to filter tree it isn't performant.
 - Lots of type checking to ensure everything will work for the filter tree
 - Moves the validation of empty searches to Query itself, which works quite nicely now for both basic and advanced.
 - Memo around the query settings since those are also expensive to rerender and already listen directly for changes on the query.

andrewkfiedler committed:
Formatting only

vinamartin commented:
🎉  Hero successful- extra kudos on fixing a lot of the crashing issues I've been seeing with query basic time.
One minor issue I saw that I also saw on my PR and have been trying to fix is that you can't "Clear All" these time properties using the "X" button **and** you also can't remove the last one. It's kind of minor but it's a pain if you only want one of the properties and have to manually click on all of the others to clear them instead of just clearing and typing what you want.
![Screen Shot 2020-10-29 at 07 17 38](https://user-images.githubusercontent.com/9013407/97586297-56266b00-19b7-11eb-9340-35098d42f4fc.png)

I think what's happening here is that we're trying to figure out what kind of filter it is on update and it doesn't work with having undefined property values. That's what I was trying to fix here: https://github.com/codice/ddf-ui/pull/437. 
I think I'm going to close out that PR in favor of writing up a patch against this one.

jordanwilking commented on ui-frontend/packages/catalog-ui-search/src/main/webapp/react-component/location/keyword.spec.js:
```diff
@@ -33,51 +33,51 @@ const { shallow } = Enzyme
-      name: 'test',
-    })
-  })
+  // it('should fetch features on select', (done) => {
```
✏️ Should we just delete all of these tests if we don't intend to replace them now?

andrewkfiedler replied:
I'm not sure to be honest, I left them commented out as a guide for what we should try and fix.  They've been not running for so long that I'm not entirely sure how relevant they are, but they probably have good ideas of what we should be testing in the components (the components have just changed too much). 

I originally had wanted to add some tests to ensure the filter builder didn't regress, but ran into issues just trying to render it.  Didn't have time to chase them down, but thought I'd include the change to make it possible to turn tests back on.

jordanwilking replied:
We could always check an earlier commit to reference these, but it's not a big deal me 🤷. I'm sure we'll revisit these as we increase our testing efforts soon.

andrewkfiedler commented:
@vinamartin nice catch, I think I can do something similar to what I did for types (I had to make a fake '*' 😭 )

andrewkfiedler committed:
Add ability to leave date properties blank on basic component

 - This introduces some complexity, which removeInvalidFilters addresses slightly, but we may want to reconsider how we display the basic component.  Since we try to be a faithful representation of the current filter tree, blanks for things like properties cause us to need to do complex things like this.

andrewkfiedler committed:
Adds ability to press enter to run search in advanced text fields

 - A special class is on these fields, so if components using the filter builder want to do some special action when enter is pressed, they now can.  See filter-builder and result-filter for examples.  They run the search when enter is pressed on a contained text field.

andrewkfiedler commented:
@vinamartin could you rehero this?  I pushed up a fix for the issue you saw, and added the ability to press enter to run a search (if within a string field on advanced search).

vinamartin commented:
Manually verified ✨ 
Looks good to me, merging
